scripts/ai/AIDriveStrategyChopperCourse.lua

```diff
@@ -0,0 +1,782 @@
+        local tentativeRendezvousWaypointIx = self:findBestWaypointToUnload(self.waypointIxWhenCallUnloader, true)
+        if not tentativeRendezvousWaypointIx then
+            self:debug('callUnloaderWhenNeeded: can\'t find a good waypoint to meet the unloader')
+            return
+        end
+        bestUnloader, bestEte = self:findUnloader(nil, self.course:getWaypoint(tentativeRendezvousWaypointIx))
+        -- getSpeedLimit() may return math.huge (inf), when turning for example, not sure why, and that throws off
+        -- our ETE calculation
+        if bestUnloader and self.vehicle:getSpeedLimit(true) < 100 then
+            local dToUnloadWaypoint = self.course:getDistanceBetweenWaypoints(tentativeRendezvousWaypointIx,
+                    self.course:getCurrentWaypointIx())
+            local myEte = dToUnloadWaypoint / (self.vehicle:getSpeedLimit(true) / 3.6)
+            self:debug('callUnloaderWhenNeeded: best unloader ETE at waypoint %d %.1fs, my ETE %.1fs',
+                    tentativeRendezvousWaypointIx, bestEte, myEte)
+            if bestEte - 5 > myEte then
+                -- I'll be at the rendezvous a lot earlier than the unloader which will almost certainly result in the
+                -- cancellation of the rendezvous.
+                -- So, set something up further away, with better chances,
+                -- using the unloader's ETE, knowing that 1) that ETE is for the current rendezvous point, 2) there
+                -- may be another unloader selected for that waypoint
+                local dToTentativeRendezvousWaypoint = bestEte * (self.vehicle:getSpeedLimit(true) / 3.6)
+                self:debug('callUnloaderWhenNeeded: too close to rendezvous waypoint, trying move it %.1fm',
+                        dToTentativeRendezvousWaypoint)
+                tentativeRendezvousWaypointIx = self.course:getNextWaypointIxWithinDistance(
+                        self.course:getCurrentWaypointIx(), dToTentativeRendezvousWaypoint)
+                if tentativeRendezvousWaypointIx then
+                    bestUnloader, bestEte = self:findUnloader(nil, self.course:getWaypoint(tentativeRendezvousWaypointIx))
+                    if bestUnloader then
+                        self:callUnloader(bestUnloader, tentativeRendezvousWaypointIx, bestEte)
+                    end
+                else
+                    self:debug('callUnloaderWhenNeeded: still can\'t find a good waypoint to meet the unloader')
+                end
+            elseif bestEte + 5 > myEte then
+                -- do not call too early (like minutes before we get there), only when it needs at least as
+                -- much time to get there as the combine (-5 seconds)
+                self:callUnloader(bestUnloader, tentativeRendezvousWaypointIx, bestEte)
+            end
+        end
+    end
+end
+
+-- Copied from parent altered to check for active chopper unloaders
+function AIDriveStrategyChopperCourse:findUnloader(combine, waypoint)

```

Comment:
same here, way too many common lines, need a more intelligent solution to not repeat business logic code.